File: 594.最长和谐子序列.cpp
class Solution {
public:
    //是子序列不是子数组！！！
    //用unordered_map记录每一个数出现的次数，数i只能和数i-1和数i+1组成和谐数组
    //找出map[i-1]+map[i]和map[i]和map[i+1]的最大值
    int findLHS(vector<int>& nums) {
        unordered_map<int, int> m;
        int Max = 0;
        for(auto i: nums){
            ++m[i];
            if(m[i-1]){
                Max = max(Max, m[i] + m[i-1]);
            }
            if(m[i+1]){
                Max = max(Max, m[i] + m[i+1]);
            }
        }
        return Max;
    }
};

File: 594.最长和谐子序列.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findLHS = function(nums)
{
    let maxLen = 0;
    let map = new Map();
    for (let i = 0; i < nums.length; ++i)
    {
        map.set(nums[i], map.has(nums[i]) ? map.get(nums[i]) + 1 : 1);
    }
    let arr = Array.from(map);
    arr = arr.sort((a, b) = > {
        return a[0] - b[0];
    }) for (let i = 1; i < arr.length; ++i)
    {
        if (arr[i][0] - arr[i - 1][0] == = 1)
        {
            maxLen = Math.max(maxLen, arr[i][1] + arr[i - 1][1]);
        }
    }
    return maxLen;
};

var findLHS = function(nums)
{
    let maxLen = 0;
    let map = new Map();
    for (let i = 0; i < nums.length; ++i)
    {
        map.set(nums[i], map.has(nums[i]) ? map.get(nums[i]) + 1 : 1);
        if (map.has(nums[i] - 1))
            maxLen = Math.max(maxLen, map.get(nums[i]) + map.get(nums[i] - 1));
        if (map.has(nums[i] + 1))
            maxLen = Math.max(maxLen, map.get(nums[i]) + map.get(nums[i] + 1));
    }
    return maxLen;
}